pkg/controller/machinepool/nutanixactuator.go

```diff
@@ -0,0 +1,223 @@
+		return nil, fmt.Errorf("error unmarshalling providerSpec: %v", err)
+	}
+
+	return spec, nil
+}
+
+// getRHCOSImageNameFromMasterMachine retrieves the RHCOS image name from the master machine provider spec.
+// Note: The machine image name is reliably set only in IPI installations. In UPI installations, this field may not be populated.
+func getRHCOSImageNameFromMasterMachine(masterMachine *machineapi.Machine, cd *hivev1.ClusterDeployment, logger log.FieldLogger) (string, error) {
+	providerSpec, err := decodeNutanixMachineProviderSpec(masterMachine.Spec.ProviderSpec.Value)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to decode provider spec while retrieving RHCOS image name from master machine")
+	}
+
+	if providerSpec.Image.Name != nil && *providerSpec.Image.Name != "" {
+		logger.Infof("using RHCOS image name from master machine providerSpec: %s", *providerSpec.Image.Name)
+		return *providerSpec.Image.Name, nil
+	}
+
+	return "", errors.Errorf("no RHCOS image name found in provider spec for ClusterDeployment %s/%s ", cd.Namespace, cd.Name)
+}
+
+// GenerateMachineSets satisfies the Actuator interface and will take a clusterDeployment and return a list of MachineSets
+// to sync to the remote cluster.
+func (a *NutanixActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, pool *hivev1.MachinePool, logger log.FieldLogger) ([]*machineapi.MachineSet, bool, error) {
+	if cd.Spec.ClusterMetadata == nil {
+		return nil, false, errors.New("ClusterDeployment does not have cluster metadata")
+	}
+
+	if cd.Spec.Platform.Nutanix == nil {
+		return nil, false, errors.New("ClusterDeployment is not for Nutanix")
+	}
+	if pool.Spec.Platform.Nutanix == nil {
+		return nil, false, errors.New("MachinePool is not for Nutanix")
+	}
+
+	computePool := baseMachinePool(pool)
+	dataDisks, err := a.getNutanixDataDisks(pool, logger)
+	if err != nil {
+		a.setUnsupportedConfigurationCondition(pool, logger, err)
+		return nil, false, nil
+	}
+
+	computePool.Platform.Nutanix = &installertypesnutanix.MachinePool{
+		NumCPUs:           pool.Spec.Platform.Nutanix.NumCPUs,
+		NumCoresPerSocket: pool.Spec.Platform.Nutanix.NumCoresPerSocket,
+		MemoryMiB:         pool.Spec.Platform.Nutanix.MemoryMiB,
+		OSDisk: installertypesnutanix.OSDisk{
+			DiskSizeGiB: pool.Spec.Platform.Nutanix.OSDisk.DiskSizeGiB,
+		},
+		BootType:       pool.Spec.Platform.Nutanix.BootType,
+		Project:        pool.Spec.Platform.Nutanix.Project,
+		Categories:     pool.Spec.Platform.Nutanix.Categories,
+		FailureDomains: pool.Spec.Platform.Nutanix.FailureDomains,
+		DataDisks:      dataDisks,
+		GPUs:           pool.Spec.Platform.Nutanix.GPUs,
+	}
+
+	ic := &installertypes.InstallConfig{
+		Platform: installertypes.Platform{
+			Nutanix: a.getNutanixPlatformInstallConfig(cd),
+		},
+	}
+
+	osImage, err := getRHCOSImageNameFromMasterMachine(a.masterMachine, cd, logger)
+	if err != nil {
+		a.setUnsupportedConfigurationCondition(pool, logger, err)
```

Comment:
Typically a method like this would return an error if the status update failed, and we would bubble that error up so we can requeue. Unless there's a specific reason to do otherwise, I would like to see that pattern followed here as well.